site/examples/code-highlighting.tsx

```diff
@@ -1,237 +1,516 @@
+        split: true,
+      }
+    )
+    Transforms.setNodes(
+      editor,
+      { type: CodeLineType },
+      { match: n => Element.isElement(n) && n.type === ParagraphType }
+    )
+  }
+
+  return (
+    <Button
+      data-test-id="code-block-button"
+      active
+      onMouseDown={event => {
+        event.preventDefault()
+        handleClick()
+      }}
     >
+      <Icon>code</Icon>
+    </Button>
+  )
+}
+
+const renderLeaf = (props: RenderLeafProps) => {
+  const { attributes, children, leaf } = props
+  const { text, ...rest } = leaf
+
+  return (
+    <span {...attributes} className={Object.keys(rest).join(' ')}>
       {children}
     </span>
   )
 }
 
-const initialValue: Descendant[] = [
+const useDecorate = (editor: Editor) => {
+  return useCallback(([node, path]) => {
+    if (Element.isElement(node) && node.type === CodeLineType) {
+      const ranges = editor.nodeToDecorations.get(node) || []
+      return ranges
+    }
+
+    return []
+  }, [])
+}
+
+class PrismTokenNode {
+  token: string | Prism.Token
+  parent: null | PrismTokenNode
+
+  constructor(token: string | Prism.Token, parent: null | PrismTokenNode) {
+    this.token = token
+    this.parent = parent
+  }
+}
+
+const getChildNodeToDecorations = ([block, blockPath]: NodeEntry<
+  CodeBlockElement
+>) => {
+  const nodeToDecorations = new Map<Element, Range[]>()
+
+  const text = block.children.map(line => Node.string(line)).join('\n')
+  const language = block.language
+  const tokens = Prism.tokenize(text, Prism.languages[language])
+  const normalizedTokens = normalizeTokens(tokens) // make tokens flat and grouped by line
+  const blockChildren = block.children as Element[]
+
+  for (let index = 0; index < normalizedTokens.length; index++) {
+    const tokens = normalizedTokens[index]
+    const element = blockChildren[index]
+
+    if (!nodeToDecorations.has(element)) {
+      nodeToDecorations.set(element, [])
+    }
+
+    let start = 0
+    for (const token of tokens) {
+      const length = token.content.length
+      if (!length) {
+        continue
+      }
+
+      const end = start + length
+
+      const path = [...blockPath, index, 0]
+      const range = {
+        anchor: { path, offset: start },
+        focus: { path, offset: end },
+        token: true,
+        ...Object.fromEntries(token.types.map(type => [type, true])),
+      }
+
+      nodeToDecorations.get(element)!.push(range)
+
+      start = end
+    }
+  }
+
+  return nodeToDecorations
+}
+
+// precalculate editor.nodeToDecorations map to use it inside decorate function then
+const SetNodeToDecorations = () => {
+  const editor = useSlate()
+
+  useMemo(() => {
+    const blockEntries = Array.from(
+      Editor.nodes(editor, {
+        at: [],
+        mode: 'highest',
+        match: n => Element.isElement(n) && n.type === CodeBlockType,
+      })
+    )
+
+    const nodeToDecorations = mergeMaps(
+      ...blockEntries.map(getChildNodeToDecorations)
```

Comment:
precalculate editor.nodeToDecorations map to use it inside decorate function then
I didn't find a way to achieve good performance without this precalculation